recipes/pcl/all/conanfile.py

```diff
@@ -0,0 +1,157 @@
+            return self._cmake
+
+        cmake_definitions = {
+            "CONAN_PCL_VERSION": self.version,
+            "PCL_BUILD_WITH_BOOST_DYNAMIC_LINKING_WIN32": self.options["boost"].shared
+        }
+
+        pcl_config = {
+            "BUILD_tools": self.options.with_tools,
+            "WITH_LIBUSB": False,
+            "WITH_PNG": True,
+            "WITH_QHULL": True,
+            "WITH_CUDA": self.options.with_cuda,
+            "WITH_VTK": False,
+            "WITH_PCAP": False,
+            "WITH_OPENGL": True,
+            "WITH_OPENNI": False,
+            "WITH_OPENNI2": False,
+            "WITH_ENSENSO": False,
+            "WITH_DAVIDSDK": False,
+            "WITH_DSSDK": False,
+            "WITH_RSSDK": False,
+            "PCL_SHARED_LIBS": self.options.shared,
+            "FLANN_USE_STATIC": not self.options["flann"].shared,
+            "QHULL_USE_STATIC": not self.options["qhull"].shared
+        }
+        pcl_features = {
+            "BUILD_kdtree": True,
+            "BUILD_octree": True,
+            "BUILD_search": True,
+            "BUILD_sample_consensus": True,
+            "BUILD_filters": True,
+            "BUILD_2d": True,
+            "BUILD_geometry": True,
+            "BUILD_io": True,
+            "BUILD_features": True,
+            "BUILD_ml": True,
+            "BUILD_segmentation": True,
+            "BUILD_surface": True,
+            "BUILD_module_registration": True,
+            "BUILD_module_keypoints": True,
+            "BUILD_module_tracking": True,
+            "BUILD_recognition": True,
+            "BUILD_stereo": True,
+        }
+
+        self._cmake = CMake(self)
+        self._cmake.definitions.update(cmake_definitions)
+        self._cmake.definitions.update(pcl_config)
+        self._cmake.definitions.update(pcl_features)
+        self._cmake.configure()
+        return self._cmake
+
+    def build(self):
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def _remove_vs_runtime_files(self):
+        patterns = [os.path.join(self.package_folder, "bin", pattern) for pattern in ["msvcp*.dll", "vcruntime*.dll", "concrt*.dll"]]
+        runtime_files = chain.from_iterable(glob(pattern) for pattern in patterns)
+        for runtime_file in runtime_files:
+            try:
+                os.remove(runtime_file)
+            except Exception as ex:
+                self.output.warn("Could not remove vs runtime file {}, {}".format(runtime_file, ex))
+
+    def package(self):
+        cmake = self._configure_cmake()
+        cmake.install()
+        self.copy(pattern="LICENSE.txt", dst="licenses", src=self._source_subfolder)
+        if self.settings.os == "Windows":
+            self._remove_vs_runtime_files()
+
+        tools.rmdir(os.path.join(self.package_folder, "cmake"))
+        tools.rmdir(os.path.join(self.package_folder, "share"))
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
+
+    def package_info(self):
```

Comment:
PCL creates several pkgconfig files (one per component) and one CMake config file (PCLConfig.cmake) with a lot of non namespaced imported targets: common kdtree octree search sample_consensus filters 2d geometry io features ml segmentation surface registration keypoints tracking recognition stereo

Reply:
Conan center index packages do not support custom PKG config or cmake files.

Reply:
Yes, but `package_info` should contain all necessary stuff to mimic those files through pkg_config, cmake_find_package and cmake_find_package_multi generators.
Even without using components, libs should be listed in proper order.

Reply:
here is a starter to help you (:

```python
    def package_info(self):
        self.cpp_info.names["cmake_find_package"] = "PCL"
        self.cpp_info.names["cmake_find_package_multi"] = "PCL"
        # common
        self.cpp_info.components["common"].names["pkg_info"] = "pcl_common"
        self.cpp_info.components["common"].libs = ["common"]
        # kdtree
        self.cpp_info.components["kdtree"].names["pkg_info"] = "pcl_kdtree"
        self.cpp_info.components["kdtree"].libs = ["kdtree"]
        self.cpp_info.components["kdtree"].requires = ["common"]
        ...
```

Maybe some values are wrong, didn't check exact lib names or pkgconfig files. You'll have also to properly define external dependencies of each component (and eventually system libs dependencies if any depending on os).

Reply:
Components is still marked as an experimental feature. Are there any other conan center index packages doing this?

Reply:
Yes, it's experimental yet, but it should not break, probably it should be increased. Yes, we have a lot of recipe using it (thanks to @SpaceIm): https://github.com/conan-io/conan-center-index/search?l=Python&q=components&type=Code

Reply:
Attempted to get as close as I can to PCL's target names, however it creates targets without namespaces along the lines of `PCL_SURFACE_LIBRARIES` etc. I've mimicked this with components, but conan's targets will be namespaced so `PCL::PCL_SURFACE_LIBRARIES`.

I don't see a way in conan/cmake to distinguish between a non-namespaced target called `PCL_SURFACE_LIBRARIES` and a plain cmake variable called `PCL_SURFACE_LIBRARIES` that contains the list of libraries to link to.

Reply:
Conan cmake generator, primarily, focus on the targets approach since that is considered the best practice now a days.

We can not create and variables (yet).

I think the components work done is complete for a first PR, feel free to mark this resolved!

Reply:
We have now several recipes which are able to provide non-namespaced targets. Take a look at OpenCV recipe for a complex example with targets.